File: examples/python/dataset_access_example.py
#!/usr/bin/env python3
"""Demonstrate the recommended DataSet access patterns."""

from __future__ import annotations

import dicomsdl as dicom


def main() -> int:
    ds = dicom.DataSet()

    rows = ds.ensure_dataelement("Rows", dicom.VR.US)
    rows.value = 512

    assert ds.set_value("Columns", 256)
    assert ds.set_value(0x00090030, dicom.VR.US, 16)

    preserved = ds.ensure_dataelement("Rows", dicom.VR.UL)

    print(f"Rows VR after ensure(UL): {preserved.vr}")
    print(f"Rows value: {ds.get_value('Rows')}")
    print(f"Columns value: {ds['Columns'].value}")
    print(f"Private value: {ds[0x00090030].value}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/dataset_access_example.cpp
#include <dicom.h>

#include <iostream>
#include <string_view>

using namespace dicom::literals;

int main() {
	dicom::DataSet ds;

	auto& rows = ds.ensure_dataelement("Rows"_tag, dicom::VR::US);
	rows.from_long(512);

	if (!ds.set_value("Columns"_tag, 256L)) {
		std::cerr << "Failed to assign Columns\n";
		return 1;
	}

	auto& private_elem = ds.ensure_dataelement(dicom::Tag(0x0009, 0x0030), dicom::VR::US);
	private_elem.from_long(16);

	auto& preserved = ds.ensure_dataelement("Rows"_tag, dicom::VR::UL);
	std::cout << "Rows VR after ensure(UL): " << preserved.vr().str() << "\n";
	std::cout << "Rows value: " << ds.get_value<long>("Rows"_tag, 0L) << "\n";
	std::cout << "Columns value: " << ds.get_value<long>("Columns"_tag, 0L) << "\n";
	std::cout << "StudyDescription: "
	          << ds.get_value<std::string_view>(
	                 "StudyDescription"_tag, std::string_view("<missing>"))
	          << "\n";
	std::cout << "Private value: " << private_elem.to_long().value_or(0) << "\n";

	return 0;
}
